### PR TITLE
For mct, run land with a subset of global threads if desired

### LINKED ISSUE #163: Have CLM listen to the driver for number of threads rather than OMP_NUM_THREADS
**Erik Kluzek < erik > - 2017-02-10 18:10:02 -0700**
**Bugzilla Id:** [2421](http://bugz.cgd.ucar.edu/show_bug.cgi?id=2421)
**Bugzilla CC:**  dlawren,  mvertens,  rfisher, 

This is CLM's part of Issue 864 for cime, where the components don't listen to what the driver sets the number of threads for them -- but use the env variable OMP_NUM_THREADS

https://github.com/ESMCI/cime/issues/864

This can be a problem if another component is threaded, and you don't want CLM to be. It'll end up being threaded anyway.

### PR DESCRIPTION
**Pull request by billsacks**
### Description of changes

Listen to the land thread number set in the mct driver rather than running on the full global number of threads.

This started from @jedwards4b 's PR #967 and supersedes that one. See comments there for more details.

However, I'm wondering if this is really the right way to do it: It feels to me like a better solution would be to set the number of threads in the driver rather than requiring all components to do this in their caps. The mct driver already has code to do this in `component_mod`:

```Fortran
             if (drv_threading) call seq_comm_setnthreads(comp(1)%nthreads_compid)
```

but maybe the `drv_threading` conditional needs to be removed for this to work right.

### Specific notes

Contributors other than yourself, if any: @jedwards4b 

CTSM Issues Fixed (include github issue #):
- Resolves #163 

Are answers expected to change (and if so in what way)? No

Any User Interface Changes (namelist or namelist defaults changes)? No

Testing performed, if any: None yet

### CODE REVIEW

**Comment by billsacks** [comment]
@jedwards4b and @ekluzek - as I mentioned above: After implementing this, I began to question the need for doing this in the component caps. It feels like a cleaner and more robust solution would be to have the driver itself set the number of threads before and after each call to the appropriate `comp_run` routine. As I mentioned above, it looks like there is sort of already code to do this in `component_run`, but it may not be working right because it's in a `drv_threading` conditional.

@jedwards4b what do you think about that? Is there really a need for every component to implement this logic, or can this be done in the driver?

**Comment by billsacks** [comment]
By the way, my general design motivation is: If there's something that needs to be done for every component, and if the driver knows enough to do it itself, then it should be done by the driver, rather than requiring every component to have duplicated logic in its cap.

File: src/cpl/mct/lnd_comp_mct.F90 [hunk 1]
```diff
@@ -276,6 +289,10 @@ subroutine lnd_init_mct( EClock, cdata_l, x2l_l, l2x_l, NLFilename )
     endif
 #endif
 
+#ifdef _OPENMP
```
Comment (jedwards4b) [inline src/cpl/mct/lnd_comp_mct.F90:292]: This one shouldn't really be necessary, but it won't hurt either.

**Comment by jedwards4b** [comment]
I think that it can be done in the driver.

**Comment by billsacks** [comment]
Thanks for the quick reply @jedwards4b . I guess I'll leave this open until it can be demonstrated to work just in the driver, without any changes necessary in CTSM or other component caps.

**Comment by ekluzek** [comment]
> By the way, my general design motivation is: If there's something that needs to be done for every component, and if the driver knows enough to do it itself, then it should be done by the driver, rather than requiring every component to have duplicated logic in its cap.

That makes sense as a principle. So you are saying that we should modify CIME/CMEPS so that they set the component threads in the driver for each generic component call. That would mean that the caps for each component won't have to set the threads, and wouldn't have to change if they aren't already setting the number of threads. And like you say there wouldn't be the duplicated code for setting threads in each component.

The one caveat I thought of for that, would be for components that can't handle threading, or can only handle certain number of threads. But, even then the component without OpenMP just would never invoke any threaded loops. And a component that had restrictions on the number of threads could check for that by checking the number of threads at initialization. So it's still OK. And a component that didn't have OpenMP could still add a check for openmp threads and die if it's greater than one.

So it seems like handling this in cime/cmeps would be the right thing to do.

**Comment by billsacks** [comment]
Based on input from @jedwards4b this is no longer needed.

**Pull request closed without merging**